tests/test_table.py

```diff
@@ -548,6 +643,204 @@ def f(x_, y_):
+        x1,y1 = 2.3, 1.2
+        np.testing.assert_allclose(tab2d(x1,y1), f(x1, y1), atol=1e-11, rtol=0)
+        ref_dfdx = dfdx(x1,y1)
+        ref_dfdy = dfdy(x1,y1)
+        test_dfdx, test_dfdy = tab2d.gradient(x1,y1)
+        np.testing.assert_allclose(test_dfdx, ref_dfdx, atol=1e-11, rtol=0)
+        np.testing.assert_allclose(test_dfdy, ref_dfdy, atol=1e-11, rtol=0)
+
+        # Check vectorized output
+        newx = np.linspace(0.2, 3.1, 45)
+        newy = np.linspace(0.3, 10.1, 85)
+        newyy, newxx = np.meshgrid(newy, newx)
+
+        np.testing.assert_allclose(tab2d(newxx, newyy), f(newxx, newyy), atol=1e-11, rtol=0)
+        ref_dfdx = dfdx(newxx, newyy)
+        ref_dfdy = dfdy(newxx, newyy)
+        test_dfdx, test_dfdy = tab2d.gradient(newxx, newyy)
+        np.testing.assert_allclose(test_dfdx, ref_dfdx, atol=1e-11, rtol=0)
+        np.testing.assert_allclose(test_dfdy, ref_dfdy, atol=1e-11, rtol=0)
+
+    # I think it ought to be possible to exactly interpolate the following,
+    # provided the exact derivatives are provided.  Use slightly looser tolerances,
+    # which seems reasonable since these are varying more rapidly.
+    def f6(x_, y_):
+        return y_*y_*y_
+    def df6dx(x_, y_):
+        return np.zeros_like(x_)
+    def df6dy(x_, y_):
+        return 3*y_*y_
+    def d2f6dxdy(x_, y_):
+        return np.zeros_like(x_)
+
+    def f7(x_, y_):
+        return 2*y_*y_*y_ + 3*x_*x_*x_ + 4*x_*y_*y_
+    def df7dx(x_, y_):
+        return 9*x_*x_ + 4*y_*y_
+    def df7dy(x_, y_):
+        return 6*y_*y_ + 8*x_*y_
+    def d2f7dxdy(x_, y_):
+        return 8*y_
+
+    fs = [f6, f7]
+    dfdxs = [df6dx, df7dx]
+    dfdys = [df6dy, df7dy]
+    d2fdxdys = [d2f6dxdy, d2f7dxdy]
+
+    for f, dfdx, dfdy, d2fdxdy in zip(fs, dfdxs, dfdys, d2fdxdys):
+        z = f(xx, yy)
+        tab2d = galsim.LookupTable2D(x, y, z, interpolant='spline',
+            dfdx=dfdx(xx, yy), dfdy=dfdy(xx, yy), d2fdxdy=d2fdxdy(xx, yy))
+
+        # Check single value functionality.
+        x1,y1 = 2.3, 1.2
+        np.testing.assert_allclose(tab2d(x1,y1), f(x1, y1), atol=1e-10, rtol=0)
+        ref_dfdx = dfdx(x1,y1)
+        ref_dfdy = dfdy(x1,y1)
+        test_dfdx, test_dfdy = tab2d.gradient(x1,y1)
+        np.testing.assert_allclose(test_dfdx, ref_dfdx, atol=1e-10, rtol=0)
+        np.testing.assert_allclose(test_dfdy, ref_dfdy, atol=1e-10, rtol=0)
+
+        # Check vectorized output
+        newx = np.linspace(0.2, 3.1, 45)
+        newy = np.linspace(0.3, 10.1, 85)
+        newyy, newxx = np.meshgrid(newy, newx)
+
+        np.testing.assert_allclose(tab2d(newxx, newyy), f(newxx, newyy), atol=1e-10, rtol=0)
+        ref_dfdx = dfdx(newxx, newyy)
+        ref_dfdy = dfdy(newxx, newyy)
+        test_dfdx, test_dfdy = tab2d.gradient(newxx, newyy)
+        np.testing.assert_allclose(test_dfdx, ref_dfdx, atol=1e-10, rtol=0)
+        np.testing.assert_allclose(test_dfdy, ref_dfdy, atol=1e-10, rtol=0)
```

Comment:
Maybe worth adding a test here of a function that is not an exact cubic, but which is only a function of x or y, that the results from using LookupTable2D matches what we get from the regular LookupTable of just that one dimension (to some precision).  Since the implementations of the spline is different in the two cases, this would be a non-trivial check.

Reply:
I think for this proposed test we'd actually expect different results (for better or worse) at the moment (see comments above about different types of splines).